nexus/src/external_api/http_entrypoints.rs

```diff
@@ -1689,17 +1675,21 @@ async fn disk_import_blocks_from_url(
 async fn disk_finalize_import(
     rqctx: RequestContext<Arc<ServerContext>>,
     path_params: Path<params::DiskPath>,
-    query_params: Query<params::FinalizeDisk>,
+    query_params: Query<params::OptionalProjectSelector>,
+    finalize_params: TypedBody<params::FinalizeDisk>,
 ) -> Result<HttpResponseUpdatedNoContent, HttpError> {
     let apictx = rqctx.context();
     let handler = async {
+        let opctx = crate::context::op_context_for_external_api(&rqctx).await?;
         let nexus = &apictx.nexus;
         let path = path_params.into_inner();
         let query = query_params.into_inner();
+        let params = finalize_params.into_inner();
+        let disk_selector =
+            params::DiskSelector { disk: path.disk, project: query.project };
+        let disk_lookup = nexus.disk_lookup(&opctx, disk_selector)?;
 
-        let opctx = crate::context::op_context_for_external_api(&rqctx).await?;
-
-        nexus.disk_finalize_import(&opctx, path.disk, query).await?;
+        nexus.disk_finalize_import(&opctx, &disk_lookup, &params).await?;
```

Comment:
While I was going through and cleaning up selectors I noticed this from #2418. There was a few things about this implementation that deviated from some of the V1 patterns. Of primary concern is passing selectors directly into the app layer. In general we want to resolve all selectors in the handlers and only pass lookups down to the app layer. Also inside of `FinalizeDisk` there was a query parameter that had runtime implications. Aside from selectors, we try to avoid query parameters that have side effects in post bodies. I separated out the selectors and moved the side effect (`snapshot_name`) into the post body.